src/driver/xen.rs

```diff
@@ -300,71 +328,106 @@ impl Introspectable for Xen {
-        let poll_result = poll(&mut fds, timeout.try_into().unwrap()).unwrap();
-        let mut pending_event_port = -1;
-        if poll_result == 1 {
-            pending_event_port = self.xev.xenevtchn_pending()?;
-            if pending_event_port != -1 {
-                self.xev
-                    .xenevtchn_unmask(pending_event_port.try_into().unwrap())?;
+        let mut fds: [PollFd; 1] = [self.evtchn_pollfd];
+        let event: Option<Event> = match poll(&mut fds, timeout.try_into()?)? {
+            0 => {
+                // timeout. no file descriptors were ready
+                None
             }
-        }
-        let back_ring_ptr = &mut self.back_ring;
-        let mut flag = false;
-        if poll_result > 0
-            && self.xev.get_bind_port() == pending_event_port
-            && RING_HAS_UNCONSUMED_REQUESTS!(back_ring_ptr) != 0
-        {
-            flag = true;
-            let req = self.xc.get_request(back_ring_ptr)?;
-            if req.version != VM_EVENT_INTERFACE_VERSION {
-                panic!("version mismatch");
+            -1 => {
+                // failure
+                return Err(Box::new(IoError::last_os_error()));
             }
-            let xen_event_type = (self.xc.get_event_type(req)).unwrap();
-            event_type = match xen_event_type {
-                XenEventType::Cr { cr_type, new, old } => EventType::Cr {
-                    cr_type: match cr_type {
-                        XenCr::Cr0 => CrType::Cr0,
-                        XenCr::Cr3 => CrType::Cr3,
-                        XenCr::Cr4 => CrType::Cr4,
-                    },
-                    new,
-                    old,
-                },
-                XenEventType::Msr { msr_type, value } => EventType::Msr { msr_type, value },
-                XenEventType::Breakpoint { insn_len, .. } => {
-                    EventType::Breakpoint { gpa: 0, insn_len }
+            1 => {
+                // event available
+                match self.xev.xenevtchn_pending()? {
+                    -1 => {
+                        // no event channel port is pending
+                        // TODO: Err
+                        panic!("No event channel port is pending");
+                    }
+                    pending_event_port => {
+                        let bind_port = self.xev.get_bind_port();
+                        if pending_event_port != self.xev.get_bind_port() {
+                            panic!(
+                                "Event received for invalid port {}, expected port {}",
+                                pending_event_port, bind_port
+                            );
+                        }
+                        // unmask
+                        self.xev.xenevtchn_unmask(pending_event_port.try_into()?)?;
+                    }
+                };
+                let back_ring_ptr = &mut self.back_ring;
+                match RING_HAS_UNCONSUMED_REQUESTS!(back_ring_ptr) != 0 {
```

Comment:
Pattern matching is overkill for a simple boolean check. Also, didn't we come to the conclusion that all those macros could have been simple functions?

Reply:
> Also, didn't we come to the conclusion that all those macros could have been simple functions?

Probably, but my focus right now is to make it work.
I'm sure we can improve afterwards.